File: IMSS/Codigos/1.py
import pandas as pd
import unicodedata
import matplotlib.pyplot as plt
from matplotlib.backends.backend_pdf import PdfPages
import os

# -----------------------------
# Función para quitar acentos
# -----------------------------
def quitar_acentos(texto):
    if isinstance(texto, str):
        return ''.join(c for c in unicodedata.normalize('NFKD', texto) if not unicodedata.combining(c))
    return texto

# -----------------------------
# Rutas de archivos y carpetas
# -----------------------------
archivo_original = r"C:\Users\Zainos\Downloads\Practica 3\Corregido\2024_Ago_ENCal_1erNivel_CSV.csv"
archivo_sin_acentos = r"C:\Users\Zainos\Downloads\Practica 3\Corregido\1_sin_acentos.csv"
pdf_graficas = r"C:\Users\Zainos\Downloads\Practica 3\Corregido\Graficas_exploracion.pdf"

# -----------------------------
# Leer CSV original
# -----------------------------
df = pd.read_csv(archivo_original, encoding='utf-8')

# -----------------------------
# Quitar acentos en los nombres de columnas
# -----------------------------
df.columns = [quitar_acentos(col) for col in df.columns]

# -----------------------------
# Quitar acentos en los valores de texto
# -----------------------------
for col in df.select_dtypes(include='object').columns:
    df[col] = df[col].apply(quitar_acentos)

# -----------------------------
# Guardar CSV limpio
# -----------------------------
df.to_csv(archivo_sin_acentos, index=False, encoding='utf-8')
print(f"Archivo sin acentos guardado")

# -----------------------------
# Crear resumen de columnas
# -----------------------------
resumen = []
for col in df.columns:
    resumen.append({
        "Columna": col,
        "Tipo": df[col].dtype,
        "Valores únicos": df[col].nunique(dropna=True),
        "Faltantes": df[col].isna().sum()
    })

resumen_df = pd.DataFrame(resumen)
resumen_csv = r"C:\Users\Zainos\Downloads\Practica 3\Corregido\resumen_columnas.csv"
resumen_df.to_csv(resumen_csv, index=False, encoding='utf-8')
print(f"Resumen de columnas terminado")

# -----------------------------
# Generar todas las gráficas en un solo PDF
# -----------------------------
with PdfPages(pdf_graficas) as pdf:
    for idx, col in enumerate(df.columns, start=1):
        unique_count = df[col].nunique(dropna=True)
        missing_count = df[col].isna().sum()
        
        plt.figure(figsize=(6,4))
        plt.bar(["Valores únicos", "Valores faltantes"], [unique_count, missing_count], color=['skyblue','salmon'])
        plt.title(f"Columna {idx}: {col}")
        plt.ylabel("Cantidad")
        plt.tight_layout()
        
        pdf.savefig()  # Guarda la figura en el PDF
        plt.close()    # Cierra la figura para no mostrarla ni acumular memoria

print(f"Todas las gráficas se han guardado en un solo PDF")

File: IMSS/Codigos/1.R
# 1.R — Limpieza de acentos, resumen de columnas y PDF de gráficas

# Paquetes
suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
  library(stringi)
  library(ggplot2)
})

# -----------------------------
# Rutas (ajusta según tu entorno)
# -----------------------------
archivo_original   <- "C:/Users/Zainos/Downloads/Practica 3/Corregido/2024_Ago_ENCal_1erNivel_CSV.csv"
archivo_sin_acentos <- "C:/Users/Zainos/Downloads/Practica 3/Corregido/1_sin_acentos.csv"
resumen_csv        <- "C:/Users/Zainos/Downloads/Practica 3/Corregido/resumen_columnas.csv"
pdf_graficas       <- "C:/Users/Zainos/Downloads/Practica 3/Corregido/Graficas_exploracion.pdf"

# -----------------------------
# Función para quitar acentos
# -----------------------------
quitar_acentos <- function(x) {
  if (is.character(x)) {
    return(stri_trans_general(x, "Latin-ASCII"))
  }
  x
}

# -----------------------------
# Leer CSV original
# -----------------------------
df <- readr::read_csv(archivo_original, show_col_types = FALSE)

# -----------------------------
# Quitar acentos en nombres de columnas y valores de texto
# -----------------------------
names(df) <- vapply(names(df), quitar_acentos, character(1))
char_cols <- names(df)[vapply(df, is.character, logical(1))]
for (col in char_cols) {
  df[[col]] <- quitar_acentos(df[[col]])
}

# -----------------------------
# Guardar CSV limpio
# -----------------------------
readr::write_csv(df, archivo_sin_acentos)
cat("Archivo sin acentos guardado\n")

# -----------------------------
# Resumen de columnas
# -----------------------------
resumen_df <- tibble::tibble(
  Columna        = names(df),
  Tipo           = vapply(df, function(x) class(x)[1], character(1)),
  `Valores únicos` = vapply(df, function(x) dplyr::n_distinct(x, na.rm = TRUE), integer(1)),
  Faltantes      = vapply(df, function(x) sum(is.na(x)), integer(1))
)
readr::write_csv(resumen_df, resumen_csv)
cat("Resumen de columnas terminado\n")

# -----------------------------
# PDF: gráficas por columna (valores únicos vs faltantes)
# -----------------------------
grDevices::pdf(pdf_graficas, width = 6, height = 4)
on.exit(grDevices::dev.off(), add = TRUE)

for (i in seq_along(df)) {
  colname <- names(df)[i]
  unique_count  <- dplyr::n_distinct(df[[i]], na.rm = TRUE)
  missing_count <- sum(is.na(df[[i]]))
  dat <- data.frame(Metrica = c("Valores únicos", "Valores faltantes"),
                    Cantidad = c(unique_count, missing_count))
  p <- ggplot(dat, aes(x = Metrica, y = Cantidad)) +
    geom_col() +
    ggtitle(sprintf("Columna %d: %s", i, colname)) +
    ylab("Cantidad") + xlab(NULL) +
    theme_minimal()
  print(p)
}
cat("Todas las gráficas se han guardado en un solo PDF\n")
